### PR TITLE
Replace moment by dayjs

### LINKED ISSUE #12575: Replace moment with day-js
##### **Overview of the feature request**

Moment is now in maintenance mode. For vue we have already date-fns, but unification across all frontends is desired. 
To make the migration as easy as possible we could try day.js as try.

##### **Motivation for or Use Case**

Get rid of old moment library. This would benefit the longterm development of jhipster and reduce the bundle size for angular/react a lot. 

##### **Related issues or PR**

* https://github.com/jhipster/generator-jhipster/issues/12518

-   [x] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by qlereboursBS**
The aim of this PR is to replace momentjs that is in maintenance mode, by DayJS that is the replacement solution proposed by moment.
Moment will be replaced for React, Angular and Vue.

fix #12575 
---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [x] Tests are added where necessary
-   [x] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### Moment usages
| File  | Use case                          |Comment                         | With DayJS |
|----------------|-------------------------------|-----------------------------|--------|
|core.module.ts.ejs (15)                     |import | `import * as moment from 'moment';`                        |                          `import  dayjs  from  'dayjs'` |
|core.module.ts.ejs      (105)                  |Substract date| `moment().year() - 100`                                    |                 `dayjs().subtract(100,  'year')` |
|date-utils.ts.ejs (19)                  |import | `import moment from 'moment';`                             |     `import  dayjs  from  'dayjs'`         |
|date-utils.ts.ejs (24)   | format date            |`moment(date).format(APP_LOCAL_DATETIME_FORMAT)`           | `dayjs(date).format(APP_LOCAL_DATETIME_FORMAT)`|
|date-utils.ts.ejs (27)                      | parse date | `moment(date, APP_LOCAL_DATETIME_FORMAT_Z).toDate()`       | `dayjs(date, APP_LOCAL_DATETIME_FORMAT_Z).toDate()` and don't forget to add `dayjs.extend(customParseFormat)` |
|date-utils.ts.ejs (29)   | format                 |`moment().startOf('day').format(APP_LOCAL_DATETIME_FORMAT)`| `dayjs().startOf('day').format(APP_LOCAL_DATETIME_FORMAT)` |
|datepicker-adapter.ts.ejs (23)       | import    |`import { Moment } from 'moment';`                         |    There's a `Dayjs` type but we're not sure its exported. |                                            |
|datepicker-adapter.ts.ejs (24)          | import   |`import * as moment from 'moment';`                        |          `import  *  as  dayjs  from  'dayjs'`               |
|datepicker-adapter.ts.ejs (27)      | Specific use         |`export class ... extends NgbDateAdapter<Moment> {`        |      There's a `Dayjs` type but we're not sure its exported. |                                           |
|datepicker-adapter.ts.ejs (28)    |  Typing           |`fromModel(date: Moment): NgbDateStruct`                   |    There's a `Dayjs` type but we're not sure its exported. |                                             |
|datepicker-adapter.ts.ejs (29)     | Check validity          |`moment.isMoment(date) && date.isValid()`                  |     `dayjs.isDayjs(date)  &&  date.isValid()`          |
|datepicker-adapter.ts.ejs (36)     |  typing          |`toModel(date: NgbDateStruct): Moment {`                   |        There's a `Dayjs` type but we're not sure its exported.                                        |
|datepicker-adapter.ts.ejs (38)    | Parse date           |`moment(date.year + '-' + date.month + '-' + date.day, 'YYYY-MM-DD')`|`dayjs(date.year  +  '-'  +  date.month  +  '-'  +  date.day,  'YYYY-MM-DD')` |
|duration.pipe.ts.ejs (3)       | import             |`import * as moment from 'moment';`                        |            `import  *  as  dayjs  from  'dayjs'`          |
|duration.pipe.ts.ejs (11)              |  Specific use (duration) |`return moment.duration(value).humanize();`                |      `return dayjs.duration(value).humanize();` |
|DurationFormat.tsx.ejs (3)      | import            |`import * as moment from 'moment';`                        |`import  *  as  dayjs  from  'dayjs'` |
|DurationFormat.tsx.ejs (20)          | Specific use (duration)        |`moment.duration(value).locale(locale).humanize();`        |     `dayjs.duration(value).locale(locale).humanize();` |                                       
|entity-management-update.component.ts.ejs (49)| import |`import * as moment from 'moment';`                        |               `import  *  as  dayjs  from  'dayjs`       |
|entity-management-update.component.ts.ejs (184) | Get current date |`const today = moment().startOf('day');`                   |`const  today  =  dayjs().startOf('day');` |
|entity-management-update.component.ts.ejs (299)| Parse date | `moment(this.editForm.get(['<%= fieldName %>'])!.value, DATE_TIME_FORMAT)`| `dayjs(this.editForm.get(['<%=  fieldName  %>'])!.value,  DATE_TIME_FORMAT)` |
|entity-management.service.spec.ts.ejs (27)    | import | `import * as moment from 'moment';`                                          | `import  *  as  dayjs  from  'dayjs'` |
|entity-management.service.spec.ts.ejs (44)  |  typing   |`moment.Moment`                                            |There's a `Dayjs` type but we're not sure its exported. |
|entity-management.service.spec.ts.ejs (58)  | Current date   |`currentDate = moment();`                                                                  | `currentDate  =  dayjs();` |
|entity.model.ts.ejs - React (26)    |  typing          |`import { Moment } from 'moment';`                                                  | There's a `Dayjs` type but we're not sure its exported. |
|entity.model.ts.ejs - Angular (26)  | typing          |`import { Moment } from 'moment';`                                                      | There's a `Dayjs` type but we're not sure its exported. |
|entity.service.ts.ejs - Angular (30)   |   import      |`import * as moment from 'moment';`                                               | `import  *  as  dayjs  from  'dayjs';` | 
|entity.service.ts.ejs - Angular (118)   |  parse date      |`moment(res.body.<%= fields[idx].fieldName %>)`            | `dayjs(res.body.<%=  fields[idx].fieldName  %>)` |
|entity.service.ts.ejs - Angular (130)   |  parse date      |`moment(<%= entityInstance %>.<%= fields[idx].fieldName %>)`| `dayjs(<%=  entityInstance  %>.<%=  fields[idx].fieldName  %>)`| 
|generator-base-private.js (370)| Specific use          |Selecting locales to keep in webpack.prod.js of Angular           | DayJS doesn't work the same way because it doesn't have a webpack plugin to whitelist languages. By default only specified languages will be loaded. We'll create a file like dayjs.config.js that will contain as much `import 'dayjs/locale/fr'` as there are languages. |
|generator-base-private.js (406)          |Specific use            |Selecting locales to keep in webpack.prod.js of React           |  Removed according to the previous comment |
|generator-base.js (567)          |Specific use| The function `getMomentLocaleId()` returns the given language id. Its using the field `momentLocaleId` from `LANGUAGES` of `generator-constants` because some specific languages have specific language id, i.e: `hy-am` for Armenian | Replace `momentLocaleId` by `dateLibLocaleId` and check if [the list of localeId are the same on DayJS than on moment](https://github.com/iamkun/dayjs/tree/dev/src/locale) |
| package.json | Specific use | usage of `moment` and `moment-locales-webpack-plugin`| It will be removed because there's no need of a webpack plugin with DayJS |
| webpack.prod.js.ejs | Specific use | Uses `moment-locales-webpack-plugin` to load the locales | It will be removed because there's no need of a webpack plugin with DayJS |

### Date fns usage
| File  | Kind of use                          |Comment                         | With DayJS |
|----------------|-------------------------------|-----------------------------|------|
| entity-update.component.spec.ts.ejs (35) | format date | import `format`, to format date with `DATE_TIME_LONG_FORMAT` format (line 133) | `import * as dayjs from 'dayjs';` then `dayjs.format(date, DATE_TIME_LONG_FORMAT)` |
| entity-update.component.spec.ts.ejs (36) | parse date | import `parseISO` unused | Remove import because it's not used |
| entity-update.component.ts.ejs (49) | format date | format date with `DATE_TIME_LONG_FORMAT` format (line 231) | replace import and then replace `format(date, DATE_TIME_LONG_FORMAT)` with `dayjs(date).format(DATE_TIME_LONG_FORMAT)` |
| entity-update.component.ts.ejs (50) | check validity | check if date is valid (line 230) | Remove the import and replace `isValid(date)` by `dayjs(date).isValid()`
| entity-update.component.ts.ejs (51) | parse date | parse date with `DATE_TIME_LONG_FORMAT` format (line 238, 246) | remove import and replace `parse(event.target.value, DATE_TIME_LONG_FORMAT, new Date())` by `dayjs(event.target.value, DATE_TIME_LONG_FORMAT)` |
| entity-update.component.spec.ts.ejs (36) | parse date | import `parseISO` unused | Remove  unused import |
| entity.service.spec.ts.ejs (26) | format date | format date with format `DATE_FORMAT` or `DATE_TIME_FORMAT` | Replace all `format(a, b)` by `dayjs(a).format(b)` |
| filters.ts.ejs (2) | format date | format date with format `DATE_TIME_FORMAT` | replace `format(new Date(value), DATE_TIME_FORMAT)` by `dayjs(new Date(value)).format(DATE_TIME_FORMAT)`|
| filters.ts.ejs (2) | parse date | parsing date | replace `parseISO(value)` by `dayjs(value)` |
| filters.ts.ejs (2) | format duration | formats a duration: https://date-fns.org/v2.15.0/docs/formatDuration | `dayjs.duration(value).humanize();` |
| filters.ts.ejs (3) | parse duration | Import of duration-fns to parse a duration (line 27) | Removed the import and use the previous line |

## Functionnal usages (checked if tested)

- [x] `date-utils.ts.ejs#convertDateTimeFromServer`: Only for React, aims to convert a date from server to the client
- [x] `date-utils.ts.ejs#convertDateTimeToServer`: Only for React, aims to convert a date from client to server
- [x] `date-utils.ts.ejs#displayDefaultDateTime`: Only for React, aims to get the default date time for inputs
- [x] `core.module.ts.ejs`: Only for Angular, aims to set the datePicker's config (`NgbDatepickerConfig`), to set minDate at -100 years from today
- [x] `datepicker-adapter.ts.ejs`: Only for Angular, it's used to convert date to map date formats between boostrap datepicker component and the model
- [x] `duration.pipe.ts.ejs`: Only for Angular, define the behavior of the `duration` pipe (`| duration`), used in entity-management and entity-manage-details screens
- [x] `DurationFormat.tsx.ejs`: Only for Angular, aims to format a duration
- [x] `entity-management-update.component.ts.ejs` Only for Angular, aims to create the data from the form value, before saving, on the entity screens
- [x] `entity-management.service.spec.ts.ejs`: Only for Angular, initializes a date for testing purpose
- [x] `entity.service.ts.ejs#convertDateArrayFromServer`: Only for Angular, in entity's service, creates a dayjs date from server when working with a list of entity
- [x] `entity-update.component.spec.ts.ejs`: Only for Vue, when testing entity update screen, import vuejs only if there's a date field, and do assertions
- [x] `entity-update.component.ts.ejs`: Only for Vue, in entity update screen, importing dayjs only if there's a date field in the entity
- [x] `entity.service.spec.ts.ejs`: Only for Vue, testing the service
- [x] `entity.service.ts.ejs#convertDateFromServer`: Only for Vue, in entity's service, creates a dayjs date from server 
- [x] `filters.ts.ejs#formatDate`: Only for Vue, this filter is used in multiple screens like admin/tracker, user-management, entity-details
- [x] `filters.ts.ejs#formatMillis`: Only for Vue, this filter is only used in admin/metrics screen
- [x] `filters.ts.ejs#duration`: Only for Vue, this filter is used in entity.vue screen, entity-details.vue

## What's left to do?

- [ ] Check that dayjs's localId are the same than moment's ones
- [ ] Check that the dates adapt where necessary when language changes
- [ ] Read PR changes and clean before submitting

### CODE REVIEW

**Comment by atomfrede** [comment]
Great work! Can you do comparison with regards to the bundle size for angular (with the webpack bundle analyzer plugin)? I am curious how much smaller it gets

File: generators/client/templates/angular/src/main/webapp/app/core/date/datepicker-adapter.ts.ejs
Comment (DanielFran) [inline generators/client/templates/angular/src/main/webapp/app/core/date/datepicker-adapter.ts.ejs:36]: https://github.com/ng-bootstrap/ng-bootstrap/issues/1544 has been fixed in ng-bootstrap 6.0.1, so probably "! after null can be removed"?

Comment (qlereboursBS) [reply generators/client/templates/angular/src/main/webapp/app/core/date/datepicker-adapter.ts.ejs]: I removed it, it's ok now, thanks

**Comment by qlereboursBS** [comment]
> Great work! Can you do comparison with regards to the bundle size for angular (with the webpack bundle analyzer plugin)? I am curious how much smaller it gets

Thanks! :)
After generating a project with the generator on the `main` branch, the bundle size is **1.06 MB**.
When generating a project with the generator on our branch (up to date with main), the bundle size is **1.01MB**

image
image

File: generators/entity-client/templates/vue/src/test/javascript/spec/app/entities/entity.service.spec.ts.ejs
Comment (qmonmert) [inline generators/entity-client/templates/vue/src/test/javascript/spec/app/entities/entity.service.spec.ts.ejs:26]: remove tab

File: generators/client/templates/react/src/main/webapp/app/config/dayjs.ts.ejs
Comment (qmonmert) [inline generators/client/templates/react/src/main/webapp/app/config/dayjs.ts.ejs:1]: I don't know if it's better to add semicolon at the end of the lines 🤔

Comment (qlereboursBS) [reply generators/client/templates/react/src/main/webapp/app/config/dayjs.ts.ejs]: Yes, I think that the convention in JHipster is to use it, you're totally right to notice this, I just added it where necessary

File: generators/entity-client/templates/vue/src/test/javascript/spec/app/entities/entity.service.spec.ts.ejs
Comment (qlereboursBS) [reply generators/entity-client/templates/vue/src/test/javascript/spec/app/entities/entity.service.spec.ts.ejs]: Done

File: generators/client/templates/angular/package.json.ejs
Comment (qlereboursBS) [reply generators/client/templates/angular/package.json.ejs]: I added this version and edited the import of the duration plugin that has been changed in the 1.9.2

**Comment by atomfrede** [comment]
@qlereboursBS Can you rebase and fix the conflict? Webflux tests are passing on main now.

**Comment by qlereboursBS** [comment]
> @qlereboursBS Can you rebase and fix the conflict? Webflux tests are passing on main now.

@atomfrede I fixed the conflicts and squashed

**Review by atomfrede — APPROVED** [review]

File: generators/client/templates/react/package.json.ejs
Comment (mshima) [inline generators/client/templates/react/package.json.ejs:38]: Move the version to package.json, so it will be managed by dependabot.

Comment (qlereboursBS) [reply generators/client/templates/react/package.json.ejs:38]: In which package.json should I put it? Should I completely remove it from from the other package.json.ejs, and how will it work?
Sorry, I would need explainations here :D

File: generators/client/templates/angular/package.json.ejs
Comment (mshima) [inline generators/client/templates/angular/package.json.ejs:47]:
```suggestion
    "dayjs": "VERSION_MANAGED_BY_CLIENT_ANGULAR",
```

And add dayjs to package.json in the same folder.
Should be done at all frameworks

File: generators/client/templates/react/package.json.ejs
Comment (mshima) [reply generators/client/templates/react/package.json.ejs:38]: Moved to angular.

File: generators/client/templates/angular/package.json.ejs
Comment (qlereboursBS) [reply generators/client/templates/angular/package.json.ejs]: Ok, I understand how it works now, so I pushed a new version, but I used the `VERSION_MANAGED_BY_CLIENT_COMMON` and I put the version in the common package.json so that all versions are managed at the same time

File: generators/client/templates/common/package.json
Comment (mshima) [inline generators/client/templates/common/package.json:3]: Should add to dependencies instead of devDependencies.

File: generators/client/templates/angular/package.json.ejs
Comment (mshima) [reply generators/client/templates/angular/package.json.ejs]: Great.

File: generators/client/templates/common/package.json
Comment (qlereboursBS) [reply generators/client/templates/common/package.json:3]: Oops, sorry

Comment (qlereboursBS) [reply generators/client/templates/common/package.json]: It's fixed

**Comment by qlereboursBS** [comment]
@mshima or @atomfrede if you have the time, could you do a final review or merge please?

**Comment by mshima** [comment]
@qlereboursBS I didn't had time to create a project with this branch (3 projects actually), otherwise it looks good to me.
Since every test is passing and dropping moment is a must to me, I will trust you and merge tomorrow if there is no objections.

**Comment by qlereboursBS** [comment]
> @qlereboursBS I didn't had time to create a project with this branch (3 projects actually), otherwise it looks good to me.
> Since every test is passing and dropping moment is a must to me, I will trust you and merge tomorrow if there is no objections.

ok @mshima let me check one last time before merging then, I'll test with both frontend clients today or tommorow and I'll let you know if it's ok to be merged

**Comment by mshima** [comment]
@qlereboursBS if you agree, I will merge and validate post merged.

We should test if https://github.com/jhipster/generator-jhipster/issues/12465 is fixed too.

**Comment by qlereboursBS** [comment]
> @qlereboursBS if you agree, I will merge and validate post merged.
> 
> We should test if #12465 is fixed too.

@mshima I just had the time to test Angular and everything is ok, but you can go, we can merge and I'll finish validating tomorrow morning for Vue and React
I'll also tell you if the #12465 is fixed, but I think so :)

**Comment by qlereboursBS** [comment]
I checked Vue and React and everything is ok :heavy_check_mark:

**Comment by WahlbergRu** [comment]
Angular 

ERROR in src/main/webapp/app/shared/util/datepicker-adapter.ts:21:19 - error TS2349: This expression is not callable.
  Type 'typeof moment' has no call signatures.

21     return date ? moment(date.year + '-' + date.month + '-' + date.day, 'YYYY-MM-DD') : null!;
                     ~~~~~~

**Pull request merged**
